### PR TITLE
Don't display the editor form if there's no editor field

### PR DESCRIPTION
**Pull request by kienstra**
#### Changes
* If there are only Inspector fields, this doesn't display the editor form:
the-ins
* Branched off https://github.com/studiopress/genesis-custom-blocks/pull/63. That accounts for most of this PR's diff. Here's the [diff](https://github.com/studiopress/genesis-custom-blocks/compare/6a2f8f692c3d34d61c7ac0271dd05213985c6a04...834cbd4c3e4df4cf4cbc40cf8b35fc8e986f0a29) for this feature only.

#### Testing instructions
1. `/wp-admin` > Custom Blocks > Create New
2. Add a field with the location of 'Inspector'
3. Publish the block
4. Create a new post
5. Add the block
6. Expected: the `<ServerSideRender>` displays, as shows in the `.gif` above. The block form never displays.

### CODE REVIEW

**Comment by lukecarbis** [comment]
Nice! I won't have time to review this until next week unfortunately.

**Comment by kienstra** [comment]
Hi @lukecarbis,
All good! 

Do you agree with the idea of this PR (without deep-diving on the code, of course)?

**Comment by lukecarbis** [comment]
Yes!

File: js/src/block-editor/controls/text.js [hunk 1]
```diff
@@ -1,3 +1,8 @@
+/**
+ * External dependencies
+ */
+import * as React from 'react';
```
Comment (lukecarbis) [inline js/src/block-editor/controls/text.js:4]: Hey! How is this useful? Noticed you included it in all the controls.

File: js/src/block-editor/controls/image.js [hunk 2]
```diff
@@ -1,11 +1,117 @@
+/**
```
Comment (lukecarbis) [inline js/src/block-editor/controls/image.js:1]: Why is this moved from components to controls?

File: js/src/block-editor/controls/color.js [hunk 3]
```diff
@@ -76,11 +33,33 @@ const GcbColorControl = ( props ) => {
...
+					onClick={ () => setIsVisible( true ) }
+				/>
+			</BaseControl>
+			{ isVisible
+				? (
+					<Popover
+						onClick={ ( event ) => event.stopPropagation() }
+						onClose={ () => setIsVisible( false ) }
+					>
+						<ColorPicker
```
Comment (lukecarbis) [inline js/src/block-editor/controls/color.js:54]: Maybe not for this PR, but could we use the default theme colours here by pulling them via:
`wp.data.select( 'core/block-editor' ).getSettings().colors`?

File: js/src/block-editor/hooks/useImage.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		imageAlt,
+		imageSrc,
+		isUploading,
+		onSelect,
+		setIsUploading,
+		uploadFiles,
+	};
+};
+
+export default useImage;
```
Comment (lukecarbis) [inline js/src/block-editor/hooks/useImage.js:119]: What does this do?

File: js/src/edit-block/components/field.js [hunk 5]
```diff
@@ -2,7 +2,7 @@
  * External dependencies
  */
 import * as React from 'react';
-import className from 'classnames';
+import classNames from 'classnames';
```
Comment (lukecarbis) [inline js/src/edit-block/components/field.js:5]: Good pickup!

File: js/src/common/constants/index.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+export const ALTERNATE_LOCATION = 'inspector';
+export const DEFAULT_LOCATION = 'editor';
+export const LOCATIONS = [ DEFAULT_LOCATION, 'inspector' ];
```
Comment (lukecarbis) [inline js/src/common/constants/index.js]: I'm not sure I like this. 🤔 

What's the purpose of abstracting "locations"? Using "Alternate" implies there could only be two locations, which is true… so why not just call them Editor and Inspector?

**Review by lukecarbis — APPROVED** [review]
Hey @kienstra! What a great improvement! There are obviously lots of changes here, so I've asked a few clarifying questions in my review, but your work is infuriatingly immaculate as always.

Obviously I've tested the patch as well, and it all works perfectly on my end. Well done! 🎉

File: js/src/block-editor/controls/color.js (hunk 3 above)
Comment (kienstra) [reply js/src/block-editor/controls/color.js:54]: Oh interesting, good point

File: js/src/edit-block/components/field.js (hunk 5 above)
Comment (kienstra) [reply js/src/edit-block/components/field.js:5]: Haha, thanks!

File: js/src/block-editor/controls/text.js (hunk 1 above)
Comment (kienstra) [reply js/src/block-editor/controls/text.js:4]: Yeah, it's a little dirty having this in every React file.

But it lets VS Code do type checking. This has revealed some incorrect props, like [here](https://github.com/studiopress/genesis-custom-blocks/pull/63/files#diff-3123e4746cd8f36584d725dcc80b1ef6041d37a82735bc5ef6a1b71bbb50899aL11)

File: js/src/block-editor/controls/image.js (hunk 2 above)
Comment (kienstra) [reply js/src/block-editor/controls/image.js:1]: Before, this file was just a wrapper for `controls/components/image.js`:

https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/block-editor/controls/image.js#L1-L12

So I though it'd be good to just move that file into this.

File: js/src/common/constants/index.js (hunk 6 above)
Comment (kienstra) [reply js/src/common/constants/index.js]: Yeah, that's a good point.

There are a few places where it's good to know that it's a default location, like line 215:

https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/edit-block/hooks/useField.js#L204-L217

https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/edit-block/components/editor.js#L84

But there might be more places where `DEFAULT_LOCATION` is more confusing than `EDITOR_LOCATION`

I could change the constant names if you'd like. As long as the value of the constant stays the same, it'll work the same whether it's named `EDITOR_LOCATION` or `DEFAULT_LOCATION`

File: js/src/block-editor/hooks/useImage.js (hunk 4 above)
Comment (kienstra) [reply js/src/block-editor/hooks/useImage.js:119]: It's handles the business logic for the Image control that used to be in [higher-order-components](https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/block-editor/components/image.js#L14-L80 like `withSelect()` and `withState()`.

**Comment by kienstra** [comment]
> There are obviously lots of changes here, so I've asked a few clarifying questions in my review, but your work is infuriatingly immaculate as always.

Haha, thanks so much!

File: js/src/block-editor/controls/image.js (hunk 2 above)
Comment (kienstra) [reply js/src/block-editor/controls/image.js:1]: Normally, this refactoring of `<Image>` would be way out of scope. 

But I noticed that the Image control didn't display the image in the 'Editor Preview', so figured I'd work on it.

File: js/src/block-editor/hooks/useImage.js (hunk 4 above)
Comment (kienstra) [reply js/src/block-editor/hooks/useImage.js:119]: Normally, this refactoring of `<Image>` would be way out of scope. 

But I noticed that the Image control didn't display the image in the 'Editor Preview', so figured I'd work on it. I didn't notice any issue in the block editor.

File: js/src/common/constants/index.js (hunk 6 above)
Comment (lukecarbis) [reply js/src/common/constants/index.js]: Of course. Totally your call on this one. Feel free to resolve.

File: js/src/block-editor/controls/color.js (hunk 3 above)
Comment (lukecarbis) [reply js/src/block-editor/controls/color.js:54]: Feel free to resolve, as I guess it's out of scope.

**Comment by lukecarbis** [comment]
Feel free to merge when ready.

File: js/src/block-editor/controls/color.js (hunk 3 above)
Comment (kienstra) [reply js/src/block-editor/controls/color.js:54]: Sure, thanks

**Comment by kienstra** [comment]
Thanks, Luke!!! 🥳 🎂

File: js/src/common/constants/index.js (hunk 6 above)
Comment (kienstra) [reply js/src/common/constants/index.js]: If it's alright, https://github.com/studiopress/genesis-custom-blocks/pull/64/commits/60337ef885d98e5fe4a121295db889a3f79c165b renames `ALTERNATE_LOCATION` to `INSPECTOR_LOCATION`, and replaces 2 uses of `DEFAULT_LOCATION` with `EDITOR_LOCATION`. 

Good call on that, default and alternate don't really describe them.
